app/util/indoorUtils.js

```
@@ -6,3 +13,158 @@ export function subwayTransferUsesSameStation(prevLeg, nextLeg) {
+  }
+  // Indoor routing starts from an entrance if the leg ends in the subway.
+  if (
+    entranceObject &&
+    ((leg.mode === 'WALK' && nextLeg?.mode === 'SUBWAY') ||
+      leg.to.stop?.vehicleMode === 'SUBWAY')
+  ) {
+    return IndoorRouteLegType.StepsAfterEntranceInside;
+  }
+  return IndoorRouteLegType.NoStepsInside;
+}
+
+export function getIndoorSteps(previousLeg, leg, nextLeg) {
+  const entranceIndex = getEntranceStepIndex(previousLeg, leg);
+  if (!entranceIndex) {
+    return [];
+  }
+  const indoorRouteLegType = getIndoorRouteLegType(previousLeg, leg, nextLeg);
+  if (indoorRouteLegType === IndoorRouteLegType.StepsBeforeEntranceInside) {
+    return leg.steps.slice(0, entranceIndex + 1);
+  }
+  if (indoorRouteLegType === IndoorRouteLegType.StepsAfterEntranceInside) {
+    return leg.steps.slice(entranceIndex);
+  }
+  return [];
+}
+
+export function isVerticalTransportationUse(type) {
+  return (
+    type === IndoorRouteStepType.ElevatorUse ||
+    type === IndoorRouteStepType.EscalatorUse ||
+    type === IndoorRouteStepType.StairsUse
+  );
+}
+
+export function getIndoorStepsWithVerticalTransportationUse(
+  previousLeg,
+  leg,
+  nextLeg,
+) {
+  return getIndoorSteps(previousLeg, leg, nextLeg).filter(step =>
+    // eslint-disable-next-line no-underscore-dangle
+    isVerticalTransportationUse(step?.feature?.__typename),
+  );
+}
+
+export function getIndoorRouteTranslationId(
+  type,
+  verticalDirection,
+  toLevelName,
+) {
+  if (type === IndoorRouteStepType.ElevatorUse && toLevelName) {
+    return 'indoor-step-message-elevator-to-floor';
+  }
+  return `indoor-step-message-${type?.toLowerCase().replace('use', '')}${
+    verticalDirection &&
+    verticalDirection !== VerticalDirection.Unknown &&
+    type !== IndoorRouteStepType.ElevatorUse
+      ? `-${verticalDirection.toLowerCase()}`
+      : ''
+  }`;
+}
+
+export function getEntranceWheelchairAccessibility(leg) {
+  return leg?.steps?.find(
```

I'd expect leg is not null. Let the code crash if somebody does not know how to use this func.

---

agreed